pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1337 @@
+				BalanceUpdate::Decrease(native_asset_reserve),
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+			Self::update_imbalance(<HubAssetImbalance<T>>::get(), delta_imbalance)?;
+
+			Self::update_hub_asset_liquidity(&BalanceUpdate::Increase(native_asset_hub_reserve))?;
+
+			Self::update_tvl(&BalanceUpdate::Increase(
+				native_asset_tvl
+					.checked_add(stable_asset_reserve)
+					.ok_or(ArithmeticError::Overflow)?,
+			))?;
+
+			<Assets<T>>::insert(T::StableCoinAssetId::get(), stable_asset_state);
+			<Assets<T>>::insert(T::NativeAssetId::get(), native_asset_state);
+
+			<HubAssetTradability<T>>::put(Tradable::SellOnly);
+
+			Self::deposit_event(Event::TokenAdded {
+				asset_id: T::StableCoinAssetId::get(),
+				initial_amount: stable_asset_reserve,
+				initial_price: stable_asset_price,
+			});
+
+			Self::deposit_event(Event::TokenAdded {
+				asset_id: T::NativeAssetId::get(),
+				initial_amount: native_asset_reserve,
+				initial_price: native_asset_price,
+			});
+
+			Ok(())
+		}
+
+		/// Add new token to omnipool in quantity `amount` at price `initial_price`
+		///
+		/// Can be called only after pool is initialized, otherwise it returns `NoStableAssetInPool`
+		///
+		/// Position NFT token is minted for LP.
+		///
+		/// Parameters:
+		/// - `asset`: The identifier of the new asset added to the pool. Must be registered in Asset registry
+		/// - `amount`: Amount of asset added to omnipool
+		/// - `initial_price`: Initial price
+		///
+		/// Emits `TokenAdded` event when successful.
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::add_token())]
+		#[transactional]
+		pub fn add_token(
+			origin: OriginFor<T>,
+			asset: T::AssetId,
+			amount: Balance,
+			initial_price: Price,
+		) -> DispatchResult {
+			let who = T::AddTokenOrigin::ensure_origin(origin)?;
+
+			ensure!(!Assets::<T>::contains_key(asset), Error::<T>::AssetAlreadyAdded);
+
+			ensure!(T::AssetRegistry::exists(asset), Error::<T>::AssetNotRegistered);
+
+			ensure!(initial_price > FixedU128::zero(), Error::<T>::InvalidInitialAssetPrice);
+
+			let (stable_asset_reserve, stable_asset_hub_reserve) = Self::stable_asset()?;
+
+			let hub_reserve = initial_price.checked_mul_int(amount).ok_or(ArithmeticError::Overflow)?;
+
+			let asset_tvl = calculate_asset_tvl(hub_reserve, (stable_asset_reserve, stable_asset_hub_reserve))
+				.ok_or(ArithmeticError::Overflow)?;
+
+			// Initial stale of asset
+			let state = AssetState::<Balance> {
+				hub_reserve,
+				shares: amount,
+				protocol_shares: amount,
+				tvl: asset_tvl,
+				tradable: Tradable::default(),
+			};
+
+			T::Currency::transfer(asset, &who, &Self::protocol_account(), amount)?;
+
+			let lp_position = Position::<Balance, T::AssetId> {
+				asset_id: asset,
+				amount,
+				shares: amount,
+				price: initial_price.into_inner(),
+			};
+
+			let instance_id = Self::create_and_mint_position_instance(&who)?;
+
+			<Positions<T>>::insert(instance_id, lp_position);
```

Comment:
there are checks of positon insereted vs expected position.  However, it is part of another test, which might be clear.

Tests are due to refactor to follow our new testing guidelines, so this will become more clear.